plugins/janus_sip.c

```diff
@@ -4521,16 +4522,35 @@ static void *janus_sip_handler(void *data) {
 				content_type = json_string_value(content_type_text);
 
 			const char *msg_content = json_string_value(json_object_get(root, "content"));
+			char custom_headers[2048];
+			janus_sip_parse_custom_headers(root, (char *)&custom_headers, sizeof(custom_headers));
+
 			if(in_dialog_message) {
 				nua_message(session->stack->s_nh_i,
 					SIPTAG_CONTENT_TYPE_STR(content_type),
 					SIPTAG_PAYLOAD_STR(msg_content),
+					TAG_IF(strlen(custom_headers) > 0, SIPTAG_HEADER_STR(custom_headers)),
 					TAG_END());
 			} else {
-				nua_message(session->stack->s_nh_r,
+				if(session->stack->s_nh_m == NULL) {
+					janus_mutex_lock(&session->stack->smutex);
```

Comment:
I'd put the if check and related actions within the mutex, and not before: otherwise you risk race conditions where two simultaneous `message` requests create different dialogs and one is lost (and leaked).

Reply:
Fixed.